### PR TITLE
Handle replay across Optimus server restarts

### PR DESCRIPTION
**Pull request by Mryashbhardwaj**

### CODE REVIEW

File: core/scheduler/service/replay_worker.go [hunk 1]
```diff
@@ -18,8 +20,14 @@ import (
 const (
 	prefixReplayed       = "replayed"
 	replayCleanupTimeout = time.Minute
+	replaySyncMultiplyer = 3
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply core/scheduler/service/replay_worker.go:23]: thanks

File: core/scheduler/service/replay_worker.go [hunk 2]
```diff
@@ -118,18 +135,18 @@ func (w *ReplayWorker) startExecutionLoop(ctx context.Context, replayID uuid.UUI
...
 
-		// delay if not the first loop
 		executionLoopCount++
+		// delay if not the first loop iteration
 		if executionLoopCount > 1 {
 			time.Sleep(time.Duration(w.config.ExecutionIntervalInSeconds) * time.Second)
 		}
 
-		w.logger.Info("[ReplayID: %s] executing replay...", replayID)
+		w.logger.Info("[ReplayID: %s] processing replay request with ID: ", replayID)
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply core/scheduler/service/replay_worker.go:144]: right

File: core/scheduler/service/replay_worker.go [hunk 3]
```diff
@@ -146,14 +163,22 @@ func (w *ReplayWorker) startExecutionLoop(ctx context.Context, replayID uuid.UUI
...
+
 		syncedRunStatus, err := w.SyncStatus(ctx, replayWithRun, jobCron)
 		if err != nil {
 			// todo: lets not kill watchers on such errors
 			w.logger.Error("[ReplayID: %s] unable to get incoming runs: %s", replayWithRun.Replay.ID().String(), err)
 			return err
 		}
 
-		if err := w.replayRepo.UpdateReplay(ctx, replayWithRun.Replay.ID(), scheduler.ReplayStateInProgress, syncedRunStatus, ""); err != nil {
+		if err := w.replayRepo.UpdateReplayRuns(ctx, replayWithRun.Replay.ID(), syncedRunStatus); err != nil {
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply core/scheduler/service/replay_worker.go:181]: yeah you are right, it should not be hardcoded to inProgress anywhere except the start

File: internal/store/postgres/scheduler/replay_repository.go [hunk 4]
```diff
@@ -228,6 +228,47 @@ func (r ReplayRepository) GetReplayJobConfig(ctx context.Context, jobTenant tena
...
+		schedulerReplayReq, err := rr.toSchedulerReplayRequest()
+		if err != nil {
+			return nil, err
+		}
+		replayReqs = append(replayReqs, schedulerReplayReq)
+	}
+	return replayReqs, nil
+}
+
+func (r ReplayRepository) AcquireReplayRequest(ctx context.Context, replayID uuid.UUID, unhandledClassifierDuration time.Duration) error {
```
[the remark this answers is not in the document]
Comment (Mryashbhardwaj) [reply internal/store/postgres/scheduler/replay_repository.go:258]: donre

**Review by arinda-arif — APPROVED** [review]

**Pull request merged**